// noomi/dist/core/main/noomi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noomi = exports.NoomiMain = void 0;
const instancefactory_1 = require("./instancefactory");
const httprequest_1 = require("../web/httprequest");
const redisfactory_1 = require("../tools/redisfactory");
const noomierror_1 = require("../tools/noomierror");
const webconfig_1 = require("../web/webconfig");
const requestqueue_1 = require("../web/requestqueue");
const dbmanager_1 = require("../database/dbmanager");
const util_1 = require("../tools/util");
const application_1 = require("../tools/application");
const launchhookmanager_1 = require("../tools/launchhookmanager");
const tipmanager_1 = require("../tools/tipmanager");
/**
 * 框架主类
 * @remarks
 * 该类实例化可进行框架初始化，框架初始化有两种方式：
 *
 * 1. new NoomiMain(config)
 *
 * 2. noomi(config)
 */
class NoomiMain {
    /**
     * 构造器
     * 初始化一个noomi应用
     * @param port -          http端口,默认3000
     * @param configPath -    配置文件路径，默认 /config，相对于项目根目录
     * @param sslPort -       https端口，默认4000
     */
    constructor(port, configPath, sslPort) {
        this.port = port || 3000;
        this.sslPort = sslPort || 4000;
        application_1.App.configPath = configPath || '/config';
        this.init().catch(r => {
            console.error(r);
        });
    }
    /**
     * 初始化框架
     */
    async init() {
        const basePath = application_1.App.configPath;
        let iniJson = null;
        // 开始启动
        console.log('read file "noomi.json".');
        try {
            const iniStr = application_1.App.fs.readFileSync(util_1.Util.getAbsPath([basePath, 'noomi.json']), 'utf-8');
            iniJson = util_1.Util.eval(iniStr);
        }
        catch (e) {
            throw new noomierror_1.NoomiError("1001") + '\n' + e;
        }
        if (iniJson === null) {
            throw new noomierror_1.NoomiError("1001");
        }
        application_1.App.appName = iniJson['app_name'] || 'APP';
        application_1.App.language = iniJson['language'] || 'zh';
        // 开始启动
        console.log(tipmanager_1.TipManager.getTip("0100"));
        // 设置是否集群
        application_1.App.isCluster = iniJson['cluster'] === true;
        if (application_1.App.isCluster && iniJson['redis'] === undefined) {
            throw new noomierror_1.NoomiError("0600");
        }
        // redis初始化
        if (iniJson.hasOwnProperty('redis')) {
            console.log(tipmanager_1.TipManager.getTip("0101"));
            const cfg = iniJson['redis'];
            if (typeof cfg === 'object') { // 配置为对象
                redisfactory_1.RedisFactory.init(cfg);
            }
            else { // 配置为路径
                redisfactory_1.RedisFactory.parseFile(util_1.Util.getAbsPath([basePath, cfg]));
            }
            console.log(tipmanager_1.TipManager.getTip("0102"));
        }
        // web config
        if (iniJson.hasOwnProperty('web')) {
            console.log(tipmanager_1.TipManager.getTip("0103"));
            const cfg = iniJson['web'];
            if (typeof cfg === 'object') { // 配置为对象
                webconfig_1.WebConfig.init(cfg);
            }
            else { // 配置为路径
                webconfig_1.WebConfig.parseFile(util_1.Util.getAbsPath([basePath, cfg]));
            }
            console.log(tipmanager_1.TipManager.getTip("0104"));
        }
        // 数据源初始化
        if (iniJson.hasOwnProperty('database')) {
            console.log(tipmanager_1.TipManager.getTip("0111"));
            const cfg = iniJson['database'];
            if (typeof cfg === 'object') { // 配置为对象
                dbmanager_1.DBManager.init(cfg);
            }
            else { // 配置为路径
                dbmanager_1.DBManager.parseFile(util_1.Util.getAbsPath([basePath, cfg]));
            }
            console.log(tipmanager_1.TipManager.getTip("0112"));
        }
        // 实例初始化
        if (iniJson.hasOwnProperty('instance')) {
            console.log(tipmanager_1.TipManager.getTip("0105"));
            await instancefactory_1.InstanceFactory.init(iniJson['instance']).catch(r => {
                console.error(r);
            });
            console.log(tipmanager_1.TipManager.getTip("0106"));
        }
        // 启动钩子执行
        await launchhookmanager_1.LaunchHookManager.run().catch(r => {
            console.error(r);
        });
        // 如果web config 配置为only https，则不创建http server
        if (!webconfig_1.WebConfig.httpsCfg || !webconfig_1.WebConfig.httpsCfg['only_https']) {
            // http 服务器
            this.server = application_1.App.http.createServer((req, res) => {
                requestqueue_1.RequestQueue.handle(new httprequest_1.HttpRequest(req, res));
            }).listen(this.port, () => {
                // Noomi启动成功！
                console.log(tipmanager_1.TipManager.getTip("0117"));
                // Http服务器正在运行,监听端口 ${0}
                console.log(tipmanager_1.TipManager.getTip("0121", this.port));
                // 启动队列执行
            }).on('error', (err) => {
                if (err.code === 'EADDRINUSE') {
                    // 地址正被使用，重试中...
                    console.log(tipmanager_1.TipManager.getTip("0118"));
                    // 1秒后重试
                    setTimeout(() => {
                        this.server.close();
                        this.server.listen(this.port);
                    }, 1000);
                }
            }).on('clientError', (err, socket) => {
                socket.end('HTTP/1.1 400 Bad Request\r\n');
            });
        }
        // https 服务器
        if (webconfig_1.WebConfig.useHttps) {
            this.httpsServer = require('https').createServer({
                key: application_1.App.fs.readFileSync(webconfig_1.WebConfig.httpsCfg['key_file']),
                cert: application_1.App.fs.readFileSync(webconfig_1.WebConfig.httpsCfg['cert_file'])
            }, (req, res) => {
                requestqueue_1.RequestQueue.handle(new httprequest_1.HttpRequest(req, res));
            }).listen(this.sslPort, () => {
                console.log(tipmanager_1.TipManager.getTip("0122", this.sslPort));
                // 启动队列执行
            }).on('error', (err) => {
                if (err.code === 'EADDRINUSE') {
                    console.error(tipmanager_1.TipManager.getTip("0118"));
                    // 1秒后重试
                    setTimeout(() => {
                        this.httpsServer.close();
                        this.httpsServer.listen(this.sslPort);
                    }, 1000);
                }
            }).on('clientError', (err, socket) => {
                socket.end('HTTP/1.1 400 Bad Request\r\n');
            });
        }
    }
}
exports.NoomiMain = NoomiMain;
/**
 * 框架主函数
 * @param port -          http端口,默认3000
 * @param contextPath -   配置文件路径，默认 /config，相对于项目根目录
 * @param sslPort -       https端口，默认4000
 * @returns             NoomiMain对象
 */
function noomi(port, contextPath, sslPort) {
    return new NoomiMain(port, contextPath, sslPort);
}
exports.noomi = noomi;
