// noomi/dist/core/main/aop/aoppointcut.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AopPointcut = void 0;
const noomierror_1 = require("../../tools/noomierror");
const util_1 = require("../../tools/util");
/**
 * Aop 切点类
 */
class AopPointcut {
    /**
     * 构造器
     * @param id -            切点id(实例内唯一)
     * @param expressions -   表达式串数组，支持通配符'*'
     * @param clazz -         切面类
     */
    constructor(id, expressions, clazz) {
        /**
         * 正则表达式数组
         */
        this.expressions = [];
        /**
         * 通知数组
         */
        this.advices = [];
        this.id = id;
        this.aspectClazz = clazz;
        if (expressions) {
            this.addExpression(expressions);
        }
    }
    /**
     * 给切点添加通知
     * @param advice -    通知对象
     */
    addAdvice(advice) {
        this.advices.push(advice);
    }
    /**
     * 添加表达式串
     * @param expr -  表达式串或数组
     */
    addExpression(expr) {
        if (!Array.isArray(expr)) {
            this.expressions.push(util_1.Util.toReg(expr));
        }
        else {
            expr.forEach((item) => {
                if (typeof item !== 'string') {
                    throw new noomierror_1.NoomiError("2001");
                }
                this.expressions.push(util_1.Util.toReg(item));
            });
        }
    }
}
exports.AopPointcut = AopPointcut;
